### PR TITLE
core/state, trie: don't leak database writes before commit

### PR DESCRIPTION
**Pull request by karalabe**
Currently both the state db and the secure trie are leaking database writes: although they were designed to cache all updates until commit, some updates still reach the database before a commit. The problem with this is that if such a key is updated multiple times, then the database will be hit with the same write multiple times. Also in certain situations the data should never even end up in the database, but it does since an intermediate state leaks it there. This makes it impossible for the state pruning to clean it up, since it cannot find these leaked entries that never should be in the database.

One of the leaks is writing a contract data whenever we update the account: if a block contains 10 transactions to a contract, we will actually write out the contract's code 10 times. Further, if a contract is created and deleted in the same block, the code will end up in the database, whereas it should not as it's only a block-internal intermediate state.

The second leak is in the secure trie, which writes the secure pre-image of a key to the database whenever the key is updated, leading to the same scenarios as above. 

The fix to both of the above scenarios was to only write them to the database during commit. For contracts this could be done easily, dropping intermediate contract code from ending up in the database. For secure tries the fix was to cache the secure entries and store them during commit. This probably still could lead to ephemeral secure hashes ending up in the database (I'll investigate and fix in a follow up PR if that's the case), but at least we only write each key once, and not once per update.

Running the old vs. new code on 100K blocks:

```
Old: Import done in 2m8.448078245s
New: Import done in 2m6.465299777s
```

Update: for 250K blocks:

```
Old: Import done in 7m45.191423932s
New: Import done in 7m33.420245411s
```

Although the difference is not *that* big, we should note that the first 100K blocks are relatively low traffic. It would probably be much more visible on higher blocks.

---

A second commit fixes a memdb bug where we didn't copy the keys of a memdb batch when saving it for later storage and the buffer area was reused externally (i.e. when iterating over a map of entries to insert).

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Wed Jan 20 10:37:35 UTC 2016_

File: trie/secure_trie.go [hunk 1]
```diff
@@ -40,9 +40,10 @@ var secureKeyPrefix = []byte("secure-key-")
 type SecureTrie struct {
 	*Trie
 
-	hash       hash.Hash
-	secKeyBuf  []byte
-	hashKeyBuf []byte
+	hash        hash.Hash
+	hashKeyBuf  []byte
+	secKeyBuf   []byte
+	secKeyCache map[string][]byte
```
Comment (Gustav-Simonsson) [inline trie/secure_trie.go]: Could this be `map[[]byte][]byte` since keys used with this map are `[]byte` ?

Comment (karalabe) [inline trie/secure_trie.go]: No, Go only allows immutable types as keys. A byte slice is not immutable. Converting it into a string actually creates a copy of the content in the background and makes it immutable.

File: trie/secure_trie.go [hunk 2]
```diff
@@ -119,16 +123,53 @@ func (t *SecureTrie) Delete(key []byte) {
...
+}
+
+// CommitTo writes all nodes and the secure hash pre-images to the given database.
+// Nodes are stored with their sha3 hash as the key.
+//
+// Committing flushes nodes from memory. Subsequent Get calls will load nodes from
+// the trie's database. Calling code must ensure that the changes made to db are
+// written back to the trie's attached database before using the trie.
+func (t *SecureTrie) CommitTo(db DatabaseWriter) (root common.Hash, err error) {
+	if len(t.secKeyCache) > 0 {
```
Comment (Gustav-Simonsson) [inline trie/secure_trie.go]: Is this check for performance?

Comment (karalabe) [inline trie/secure_trie.go]: Yes, mostly so that if the map is empty, we don't allocate a new one. Probably doesn't matter that much.

**Comment by obscuren** [comment]
:+1: (forgot last week)

**Pull request merged**